src/autograd/Functions.cpp

```diff
@@ -0,0 +1,251 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/autograd/Variable.hpp>
+#include <af/autograd/Functions.hpp>
+
+namespace af {
+    namespace autograd {
+
+        Variable operator +(const Variable &lhs, const Variable &rhs)
+        {
+            auto result = lhs.array() + rhs.array();
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
```

Comment:
And using `std::array` is not an option. All functions need to share the same signature so they can be stored as `GradFunc_t` inside Variable.